README.md
# amrhike

A proof-of-concept for storage and querying of [harmonized Antimicrobial Resistance Genomic Analysis Results](https://github.com/pha4ge/harmonized-amr-parsers)

## Installation

Follow installation instructions for [Leiningen](https://leiningen.org/) for your system

## Usage

```bash
lein run
```

Currently, the program is designed to load a small set of harmonized AMR Genomic Analysis Result
files into a [datahike](https://github.com/replikativ/datahike) database.

It then runs the following query:

```edn
[:find
 ?sample ?tool ?gene ?contig ?start ?stop
 :where [?e :gene_symbol "catA1"]
        [?e :gene_symbol ?gene]
        [?e :sample_id ?sample]
        [?e :analysis_software_name ?tool]
        [?e :contig_id ?contig]
        [?e :start ?start]
        [?e :stop ?stop]]
```

...which essentially means "find all results where the `catA1` gene was found, and display a subset of fields associated with those results"

The result is printed in JSON format to standard output, and should look like:

```json
[ {
  "sample_id" : "SAMEA6058467",
  "analysis_software_name" : "AMRFinderPlus",
  "gene_symbol" : "catA1",
  "contig_id" : "DAAGAT010000085.1",
  "start" : 43,
  "stop" : 699
}, {
  "sample_id" : "SAMEA6058467",
  "analysis_software_name" : "AMRFinderPlus",
  "gene_symbol" : "catA1",
  "contig_id" : "DAAGAT010000041.1",
  "start" : 222,
  "stop" : 878
}, {
  "sample_id" : "SAMEA6058467",
  "analysis_software_name" : "ABRicate",
  "gene_symbol" : "catA1",
  "contig_id" : "DAAGAT010000041.1",
  "start" : 222,
  "stop" : 881
}, {
  "sample_id" : "SAMEA6058467",
  "analysis_software_name" : "ABRicate",
  "gene_symbol" : "catA1",
  "contig_id" : "DAAGAT010000085.1",
  "start" : 43,
  "stop" : 702
} ]
```

## License

Copyright © 2020 Dan Fornika

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2020-04-02
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2020-04-02
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/amrhike/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/amrhike/compare/0.1.0...0.1.1


test/amrhike/core_test.clj
(ns amrhike.core-test
  (:require [clojure.test :refer :all]
            [amrhike.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/amrhike/core.clj
(ns amrhike.core
  (:require [datahike.api :as d]
            [clojure.java.io :as io]
            [clojure.edn :as edn]
            [cheshire.core :refer :all]))

(defn load-edn
  "Load edn from an io/reader source (filename or io/resource)."
  [source]
  (try
    (with-open [r (io/reader source)]
      (edn/read (java.io.PushbackReader. r)))
    (catch java.io.IOException e
      (printf "Couldn't open '%s': %s\n" source (.getMessage e)))
    (catch RuntimeException e
      (printf "Error parsing edn file '%s': %s\n" source (.getMessage e)))))

(defn remove-nils
  [record]
  (into {} (remove (comp nil? second) record)))

(defn -main
  ""
  [& args]
  (def uri "datahike:file:///tmp/amrhike")
  (def schema (load-edn (io/resource "schema.edn" )))
  (def abricate-data (vec (map remove-nils (parse-stream (io/reader (io/resource "abricate-data.json")) true))))
  (def amrfinder-data (vec (map remove-nils (parse-stream (io/reader (io/resource "amrfinder-data.json")) true))))
  (def rgi-data (vec (map remove-nils (parse-stream (io/reader (io/resource "rgi-data.json")) true))))
  
  (d/delete-database uri)
  (d/create-database uri)
  (def conn (d/connect uri))
  (d/transact conn schema)
  
  (d/transact conn abricate-data)
  (d/transact conn amrfinder-data)
  (d/transact conn rgi-data)
  
  (def catA1-hits-query '[:find
                          ?sample ?tool ?gene ?contig ?start ?stop
                          :where
                          [?e :gene_symbol "catA1"]
                          [?e :gene_symbol ?gene]
                          [?e :sample_id ?sample]
                          [?e :analysis_software_name ?tool]
                          [?e :contig_id ?contig]
                          [?e :start ?start]
                          [?e :stop ?stop]])

  (def output-keys [:sample_id :analysis_software_name :gene_symbol :contig_id :start :stop])
  (def cat1-hits (map #(zipmap output-keys %) (d/q catA1-hits-query @conn)))
  
  (print (generate-string cat1-hits {:pretty true}))
  (d/release conn)
  (d/delete-database uri)
)

(comment
  (def amr-genomic-results [{:sample_id "SAMEA6058467"
                             :input_file_name "GCA_009625195.1_PDT000607292.1_genomic.fna"
                             :contig_id "DAAGAT010000041.1"
                             :start 222
                             :stop 881
                             :strand_orientation "+"
                             :gene_symbol "catA1"
                             :coverage_percent 100.0
                             :sequence_identity 99.85
                             :reference_database_id "ncbi"
                             :reference_database_version "35f5ea86fce565dd6861f79cbc578b7cc4c3d604"                 
                             :analysis_software_name "ABRicate",
                             :analysis_software_version "0.9.8"}
                            {:sample_id "SAMEA6058467"
                             :contig_id "DAAGAT010000041.1"
                             :start 222
                             :stop 878
                             :strand_orientation "+"
                             :gene_symbol "catA1"
                             :coverage_percent 100.0
                             :sequence_identity 100.0
                             :reference_database_id "ncbi"
                             :reference_database_version "2020-01-22.1"
                             :analysis_software_name "AMRFinderPlus"
                             :analysis_software_version "3.6.10"}])
  )
